### PR TITLE
Fix: Track current and previous cycle start times to set the right value in blob header

### PR DESCRIPTION
**Pull request by Sunjeet**
Cycle start time is set by the producer in the blob headers, using local time in milliseconds, corresponding to the write state creation. Consumers rely on this value to compute a metric for measuring data staleness- time elapsed since cycle started on the producer.

The earlier implementation always injected the start time of the current runCycle into the blob header. This was problematic for reverse deltas since the reverse delta from v1 to v0 is created in runcycle corresponding to v1 and thus contained the start time for v1 runcycle. Instead, the reverse detla header should contain the start time corresponding to v0 cycle, so that consumers can accurately compute staleness when they transition to v0 state by following reverse delta from v1 to v0.

In the reworked implementation, previous cycle start time is tracked in the write state engine and thats set on the reverse delta.

### CODE REVIEW

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java [hunk 1]
```diff
@@ -280,13 +280,17 @@ private void writeHeaders(DataOutputStream os, ProducerOptionalBlobPartConfig.Op
 
         /// write main header
         HollowBlobHeader header = new HollowBlobHeader();
-        header.setHeaderTags(stateEngine.getHeaderTags());
+        header.getHeaderTags().putAll(stateEngine.getHeaderTags());
```
Comment (Sunjeet) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java]: subtle change here to make a deep copy

File: hollow/src/main/java/com/netflix/hollow/core/util/HollowWriteStateCreator.java [hunk 2]
```diff
@@ -191,6 +193,10 @@ public void run() {
         
         writeEngine.addHeaderTags(readEngine.getHeaderTags());
         writeEngine.overrideNextStateRandomizedTag(readEngine.getCurrentRandomizedTag());
+        if (readEngine.getHeaderTag(HEADER_TAG_METRIC_CYCLE_START) != null) {
```
Comment (nayanika-u) [inline hollow/src/main/java/com/netflix/hollow/core/util/HollowWriteStateCreator.java]: Not very clear on why we use the header from read state and use that value to override the write state. Why is this needed?

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java (hunk 1 above)
Comment (nayanika-u) [reply hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java]: +1 since we are updating header tags in different places.

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java [hunk 3]
```diff
@@ -280,13 +280,17 @@ private void writeHeaders(DataOutputStream os, ProducerOptionalBlobPartConfig.Op
...
         if(isReverseDelta) {
             header.setOriginRandomizedTag(stateEngine.getNextStateRandomizedTag());
             header.setDestinationRandomizedTag(stateEngine.getPreviousStateRandomizedTag());
+            // cycle start time on reverse delta from v1 to v0 is the cycle start time of v0
+            stateEngine.getPreviousCycleStartTs().ifPresent(t -> header.setCycleStartTsTag(t));
         } else {
             header.setOriginRandomizedTag(stateEngine.getPreviousStateRandomizedTag());
             header.setDestinationRandomizedTag(stateEngine.getNextStateRandomizedTag());
+            // cycle start time on delta from v0 to v1 is the cycle start time of v1
```
Comment (nayanika-u) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java:292]: For the comment at 292 : Shouldn't this also be cycle start time of v0?

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java [hunk 4]
```diff
@@ -168,6 +171,11 @@ public void run() {
 
         previousStateRandomizedTag = readStateEngine.getCurrentRandomizedTag();
         nextStateRandomizedTag = mintNewRandomizedStateTag();
+        if (readStateEngine.getHeaderTag(HEADER_TAG_METRIC_CYCLE_START) != null) {
+            previousCycleStartTs = OptionalLong.of(Long.valueOf(
```
Comment (nayanika-u) [inline hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:175]: I understand that we keep track of previous cycle TS. In restoreFrom(), If a producer is restarted then why do we need to capture the previous cycle TS? Isn't this value redundant in this scenario?

Comment (nayanika-u) [reply hollow/src/main/java/com/netflix/hollow/core/write/HollowWriteStateEngine.java:175]: Okay, this value is assigned during reverse delta computation..

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java (hunk 1 above)
Comment (nayanika-u) [reply hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java]: Are we validating the size of header tags map in the UT?

**Review by nayanika-u — APPROVED** [review]

File: hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java (hunk 3 above)
Comment (nayanika-u) [reply hollow/src/main/java/com/netflix/hollow/core/write/HollowBlobWriter.java:292]: just bumping this one^

**Comment by Sunjeet** [comment]
Abandoned in favor of [this PR](https://github.com/Netflix/hollow/pull/564) which fixes how header tags are populated on reverse deltas. The current implementation sets the cycle start time as a header tag within runCycle, so that fix resolves this
